CMake: Add initial support

hugueskamba opened the pull request:
<!--
For more information on the requirements for pull requests, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html).

NOTE: Do not remove any of the template headings (even for optional sections) as this
template is automatically parsed. 
-->

### Summary of changes <!-- Required -->
Addition of CMake modules and input source files to configure the project and generate the scripts needed by the build systems (tested with Unix Makefiles and Ninja) to build Mbed OS as a library. Mbed OS is built as collection of all libraries supported by a given target. 

There are two main targets: `mbed-os` and `mbed-baremetal`. The rest of the components in the Mbed OS repository are targets prefixed with `mbed-`, for example `mbed-ble` and `mbed-netsockets`.

Even though this is still under development. This PR was raised as the current state does not affect builds of the OS with the traditional tools (Mbed CLI, Online compiler, Mbed Studio).

For current state of development, supported targets, supported toolchains, known issues, and build steps, see "Documentations" section below.

<!-- 
    Please provide the following information: 

    Description of the the change (what is this fixing / adding / removing?).

    Why the change is needed (if this is fixing a reported issue please summarize what
    the issue is and add the reference. E.g. Fixes #17119).

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
    
-->

#### Impact of changes <!-- Optional -->
<!-- 
    If there are any implications for users taking this change then they must be 
    provided here. For Major PR types this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->
Adding/removing source files or subdirectories requires adding or updating CMake input source files as all source files are listed in them.


#### Migration actions required <!-- Optional -->
<!-- 
    This should only be applicable in Major PR types for which this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

### Documentation <!-- Required -->
https://github.com/ARMmbed/mbed-os/blob/feature-cmake/tools/cmake/README.md
<!-- 
    Please provide details of any document updates required, including links to any
    related PRs against the docs repository.
    If no document updates are required please specify 'None', this at least tells us
    that this has been considered.
-->

----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

<!--
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front of them would change the meaning incorrectly. 
-->
    [] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [X] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results <!-- Required -->

<!--
    Provide all the information required, listing all the testing performed. For new targets please attach full test results for all supported compilers.
-->
    [] No Tests required for this change (E.g docs only update)
    [X] Covered by existing mbed-os tests (Greentea or Unittest)
    [ ] Tests / results supplied as part of this PR
    
    
----------------------------------------------------------------------------------------------------------------
### Reviewers <!-- Optional -->
<!--
    Request additional reviewers with @username or @team
-->

----------------------------------------------------------------------------------------------------------------

hugueskamba committed:
CMake: remove version number from ARM toolchain

The version number was required previously when ARM Compiler 5
and ARM Compiler 6 were both supported. There was a significant
underlying change between the two that justified having distinct
build options settinfgs. It is very unlikely that such a major change will happen
between ARM Compiler 6 and future versions. It is therefore also very unlikely
that distinct option settings for future versions will be needed.
We should therefore avoid appending version number the same way we do not
append it to GCC toolchain configuration settings.

hugueskamba committed:
CMake: Add design document

hugueskamba committed:
CMake: Intentionally untrack CMake generated files

hugueskamba committed:
CMake: Add CMake scripts

hugueskamba committed:
CMake: Add support for cmsis dir

hugueskamba committed:
CMake: Add support for components dir

hugueskamba committed:
CMake: Add support for connectivity dir

hugueskamba committed:
CMake: Add support for drivers dir

hugueskamba committed:
CMake: Add support for events dir

hugueskamba committed:
CMake: Add support for features dir

hugueskamba committed:
CMake: Add support for hal dir

hugueskamba committed:
CMake: Add support for platform dir

hugueskamba committed:
CMake: Add support for rtos dir

hugueskamba committed:
CMake: Add support for storage dir

hugueskamba committed:
CMake: Add support for K64F target

hugueskamba committed:
CMake: Add support for NRF52840_DK target

hugueskamba committed:
CMake: Add support for DISCO_L475VG_IOT01A target

hugueskamba committed:
CMake: Add support for WIO_3G target

hugueskamba committed:
CMake: Add Travis script to run a basic CI check for PRs

0xc0170 committed:
CMake design doc: update to the latest mbed-tools updates

Refactor Mbed OS CMake scripts section - details about mbed-os/cmake scripts
Add mbed-tools - how mbed-tools built an application with CMake

urutva committed:
cmake: Stringify develop.cmake

Cleanup develop.cmake to remove ARMC5 and uArm compiler flags and
stringify cmake variables.

Signed-off-by: Devaraj Ranganna <devaraj.ranganna@arm.com>

urutva committed:
cmake: Using Arm compiler from Mbed Studio

Added a command line option `-DMBEDIDE=ON` which allows using Arm
compiler shipped with Mbed Studio on local machines.

Signed-off-by: Devaraj Ranganna <devaraj.ranganna@arm.com>

hugueskamba committed:
CMake: Fix failure due to restructure of rtos and ppp directories

0xc0170 committed:
Travis: add more examples we should support

Co-authored-by: Hugues Kamba <41612201+hugueskamba@users.noreply.github.com>

hugueskamba committed:
CMake: Fix CI error due to dir restruct of rtos and storage dirs

urutva committed:
cmake: Re-order cmake script inclusions

The script `cmake/env.cmake` enables languages and in the case of Arm
compiler, `CMAKE_SYSTEM_PROCESSOR` is expected to be set before a
language is enabled. `CMAKE_SYSTEM_PROCESSOR` is set in
`cmake/core.cmake`. Therefore, `cmake/core.cmake` is included before
`cmake/env.cmake`.

Similarly, `cmake/profile.cmake` sets compiler options, hence it is
included before enabling languages in `cmake/env.cmake`.

Signed-off-by: Devaraj Ranganna <devaraj.ranganna@arm.com>

urutva committed:
cmake: Fix arm compiler build failures

Update arm compiler flags which were causing build failures.

Remove `-MMD` flag as it is not recognised by the Arm compiler.

There are no common flags between compiler, assembler and linker. The
compiler expects `-mcpu` whereas assembler and linker expects `--cpu`.

The `COMMON_FLAGS` in `cmake/toolchains/ARM.cmake` are not recognised by
assembler and linker.

Signed-off-by: Devaraj Ranganna <devaraj.ranganna@arm.com>

0xc0170 committed:
CMake env: fix copyright

0xc0170 committed:
CMake env: remove components, needs design first

0xc0170 committed:
CMake: fix missing copyright

0xc0170 committed:
CMake util: fix copyright

Patater committed:
cmake: Remove IAR support

IAR isn't supported anymore. Remove references to IAR from the CMake
build system generator scripts.

urutva committed:
cmake: Armclang - Fix assembler and linker flags

For Armclang, change compiler flag to `-mcpu` and assembler and linker
flag to `--cpu`.

Signed-off-by: Devaraj Ranganna <devaraj.ranganna@arm.com>

hugueskamba committed:
CMake: Use modern CMake techniques to set toolchain options

Use target_compile_options(), target_compile_definitions(),
and target_link_options() to set toolchain options.

hugueskamba committed:
CMake: Remove unsupported ARM toolchain support

hugueskamba committed:
CMake: Fix failure due to kvstore and cellular dirs changes

0xc0170 committed:
CMake: using CMAKE_BUILD_TYPE for choosing a profile

Select a profile using `-DCMAKE_BUILD_TYPE=<profile>`.
We support 3 profiles: Debug, Develop and Release.

hugueskamba committed:
CMake: Ensure targets creation before usage in `target_` functions

Move target creation statements before include statements. The targets
must be created before they can be used in `target_` functions.

hugueskamba committed:
CMake: Issue a fatal error if a lower version of CMake is used

rwalton-arm committed:
Remove gen_config interface library target

hugueskamba committed:
CMake: Fix build due to system_storage and netsocket dir move in master

0xc0170 committed:
CMake build type: set default Develop profile

Adding also comments for multiconfig/property set.

0xc0170 committed:
CMake profile: fix alignment

0xc0170 committed:
Travis: use build type for cmake

0xc0170 committed:
CMake profile: fix comments

rwalton-arm committed:
Add tools generated definitions and properties

0xc0170 committed:
Travis cmake: add testing all profiles

0xc0170 committed:
CMake toolchain: set optimization in profiles

0xc0170 committed:
CMake: set debug optimizations

0xc0170 committed:
CMake: set develop optimizations

0xc0170 committed:
CMake: set release optimizations

0xc0170 committed:
CMake toolchain: clear flags to be able to overwrite them

As Mbed uses different optimizations levels, we should be able to overwrite this via toolchain file.

0xc0170 committed:
CMake GCC ARM: use nosys specs for linker

Use retarged version (nosys.specs) for Mbed OS. To resolve an issue with debug/develop profiles.
Otherwise we get kill or getpid errors from linker, as we use nosys lib but no specs defined.

0xc0170 committed:
CMake: move env to toolchain file (#13369)

Toolchain sets target toolchain related settings that were in env. We do not need env cmake.
See https://cmake.org/cmake/help/v3.8/manual/cmake-toolchains.7.html for details

hugueskamba committed:
CMake: Fix failure due to events and netsocket dirs changes

hugueskamba committed:
CMake: Remove app.cmake module

This module was doing things that mbed-os and the application were supposed to do.
Moved the statements form the module to `mbed-os/CMakeLists.txt` and to
`<APPLICATION_ROOT>/CMakeLists.txt` .
Mbed OS also ensures the executable produced uses whatever name the application has set.

0xc0170 committed:
CMake GCC ARM: add -g3 to the common

Remove -g as g3 already provided

hugueskamba committed:
CMake: Remove add_executable

Instead provide functions to set the pre and post link commands.

hugueskamba committed:
CMake: Fix build failure due to nanostack drivers move

0xc0170 committed:
CMake toolchain: fix options comment

Co-authored-by: Hugues Kamba <41612201+hugueskamba@users.noreply.github.com>

0xc0170 committed:
CMake: Profile comments update

Co-authored-by: Hugues Kamba <41612201+hugueskamba@users.noreply.github.com>

hugueskamba committed:
CMake: Break down RTX CMake source file into 2 source files

hugueskamba committed:
CMake: Fix ARM toolchain compilation

Ensure the MCU core CMake module is added before the toolchain
CMake module as the latter requires CMAKE_SYSTEM_PROCESSOR
to be set when configuring the ARM toolchain.

hugueskamba committed:
CMake: Move MBED_STUDIO_ARM_COMPILER to ARM.cmake module

As its name implies, the flag is specific to the ARM toolchain.

hugueskamba committed:
CMake: Remove MCU core specific option from ARM ASM options

hugueskamba committed:
CMake: Fix detection of RTX assembly file

Access MBED_TARGET_LABELS as a global variable instead of a property.
At this point the evaluation of the property MBED_TARGET_LABELS
created in the top level CMakeLists.txt results in an empty string.

0xc0170 committed:
Update cmake/toolchain.cmake

Co-authored-by: Hugues Kamba <41612201+hugueskamba@users.noreply.github.com>

hugueskamba committed:
CMake: Do not hard code target in CPU core CMake module

Provide a function in each CPU core module to set toolchain options.
All the functions have the same interface therefore the CMake source
file including the module always calls the same function regardless
of which cpu core module is actually included.
When the `mbed-os` target is broken up into multiple libraries, a
library other than `mbed-os` might need to set the toolchain options.
This will be possible by simply including the module and calling the
function with the target and toolchain for which to set the options.

hugueskamba committed:
CMake: Remove toolchain module dependency on cpu core CMake modules

Set `CMAKE_SYSTEM_PROCESSOR` in toolchain.cmake based on
`MBED_CPU_CORE`.
This information is needed by the ARM toolchain (and possibly other
toolchains), it therefore makes sense to set in `toolchain.cmake`.

hugueskamba committed:
CMake: Do not hard code target in toolchain CMake module

Provide a function in each toolchain module to set the toolchain
specific options. All the functions have the same interface
therefore the CMake source file including the module always calls
the same function regardless of which toolchain module is actually
included.
When the `mbed-os` target is broken up into multiple libraries, a
library other than `mbed-os` might need to set the toolchain options.
This will be possible by simply including the module and calling the
function with the target for which to set the options.

hugueskamba committed:
CMake: Do not hard code target in build profile CMake module

Provide a function in each build profile module to set the toolchain
options. All the functions have the same interface therefore the
CMake source file including the module always calls the same
function regardless of which build profile module is actually
included.
When the `mbed-os` target is broken up into multiple libraries,
a library other than `mbed-os` might need to set the toolchain
options. This will be possible by simply including the module
and calling the function with the target and toolchain for which
to set the options.

hugueskamba committed:
CMake: Remove utility function dependency on mbed-os target

The utility function may be called by targets other than mbed-os
when it is broken into multiple libraries.

hugueskamba committed:
CMake: Fix build failure due to drivers and lwipstack dirs move

hugueskamba committed:
CMake: Fix build issue due to platform changes

hugueskamba committed:
CMake: Fix access to cellular drivers

hugueskamba committed:
CMake: Fix access to usb hal source file

hugueskamba committed:
CMake: Fix WIO_3G scatter file shebang to work with ARM Compiler 6

hugueskamba committed:
CMake: Minor formatting of top level CMakeLists.txt source file

hugueskamba committed:
CMake: Fix build failure due to nanostack dir move

hugueskamba committed:
CMake: Set language std per target using properties

hugueskamba committed:
CMake: Remove default build type selection from top level source file

hugueskamba committed:
CMake: Fix breakage due to BLE dir move

hugueskamba committed:
CMake: Fix failure due to BLE component dir move

hugueskamba committed:
CMake: Use `mbed_` prefix for CMake functions

hugueskamba committed:
CMake: Fix failure due to MbedTLS rdir relocation

rwalton-arm committed:
CMake: Check CMake builds CMake-capable targets in Travis

0xc0170 committed:
CMake: use response files

As Mbed OS is built as whole, we have long paths again. This is known issue with windows. To fix the paths, we need to use response files.

Ninja is special, needs to be forced to use long paths.

hugueskamba committed:
CMake: Fix build failure due to lora, cryptocell, and nfc dirs move

0xc0170 committed:
CMake: Fix include path for Cryptocell

hugueskamba committed:
CMake: Fix NRF52840_DK build failure

* Remove duplicated ARM toolchain directory
* Link with archived cryptocell 310 libraries instead
  of adding them as source files
* Provide default MBED_BOOT_STACK_SIZE definition
  in scatter file as is done for GCC_ARM linker file

hugueskamba committed:
CMake: Fix CORDIO_LL build failure

List missing CORDIO_LL source file in CMake source file.

hugueskamba committed:
CMake: Fix failures due to device_key and storage dirs move

0xc0170 committed:
cmake: fix sd component recent dir structure

0xc0170 committed:
CMake: use CMAKE_TOOLCHAIN_FILE

We should define CMAKE_TOOLCHAIN_FILE (as we are providing Mbed OS toolchain file). This fixes
also the issue we had with Unix makefiles where application would be without response files.

For more details, see https://gitlab.kitware.com/cmake/cmake/-/issues/21104

As the Mbed OS toolchain file is included by CMake, we need to have enabling language in the main CMake (can be executed
only once).

0xc0170 committed:
CMake: Fix blockdevice components source and include paths

This file paths were updated on master and this was broken after
rebasing.

hugueskamba committed:
CMake: Fix build failure due to PSA dir move

hugueskamba committed:
CMake: List missing files in sal-stack-nanostack

hugueskamba committed:
CMake: Add support for K66F

hugueskamba committed:
CMake: Correct file listing for PSA

hugueskamba committed:
CMake: Fix selection of scatter file and startup file for DISCO_L475_IOT01A

0xc0170 committed:
CMake ARMClang6: use armclang for asm

Fixes the issue of armasm, passing -DMACRO={0,0} - fails to parse.

We use armclang frontend but passing -masm=armasm.

0xc0170 committed:
CMake cores: fix flag  mpcu for asm

hugueskamba committed:
CMake: Fix build failure due to CMSIS restructure

0xc0170 committed:
CMake ARMClang: use -masm=auto

Auto mode for ASM files. There might be rare cases where this fails, we will fix
them by requesting the component to use gnu syntax for assembly files.

hugueskamba committed:
CMake: Fix build due to hal dir move

0xc0170 committed:
CMake: 3.18.2 minimum version set

There are fixes in the latest version we need for ARMClang.

hugueskamba committed:
CMake: mbed_add_cmake_directory_if_labels checks presence of CMakeLists.txt before adding subdirectory

0xc0170 committed:
CMake: Add information about the current state of support (#13551)

hugueskamba committed:
CMake: Fix build failure due to mbed-trace, mbed-client-randlib, and nanostack dir changes

hugueskamba committed:
CMake: Clean up platform

donatieng commented on cmake/README.md:
```diff
@@ -0,0 +1,55 @@
+## CMake for Mbed OS
+
+Until further notice, CMake for Mbed OS is still in development. 
```
Could you add a blurb about:
* what this is
* why we're doing it
* what's the plan (what's there, what's being worked on)

0xc0170 replied:
We will add more details

hugueskamba replied:
@donatieng 
Would a link to [this](https://os.mbed.com/blog/entry/Introducing-the-new-Mbed-Tools/) be satisfactory?

donatieng replied:
A link would be good, but additionally I think we need a concise paragraph, focused on the CMake work.

hugueskamba replied:
Please see: https://github.com/ARMmbed/mbed-os/pull/13589

donatieng commented on docs/design-documents/tools/cmake.md:
```diff
@@ -0,0 +1,116 @@
+# CMake Mbed OS
```
Same a pointer to the background and why/what would be helpful

0xc0170 replied:
We will update this

donatieng commented on platform/mbed-trace/CMakeLists.txt:
```diff
@@ -1,17 +1,23 @@
 
-cmake_minimum_required (VERSION 2.8)
-SET(CMAKE_SYSTEM_NAME Generic)
+if(DEFINED TARGET_LIKE_X86_LINUX_NATIVE)
```
Why make it specific to the x86 architecture? How about running on an Arm laptop/desktop/server?

0xc0170 replied:
This one comes from external lib, it was consolidated, see https://github.com/ARMmbed/mbed-trace/blob/master/source/CMakeLists.txt. It looks like we changed it but we only consolidated the cmakelists as it was split into multiple. It might be better to just revert this and let it as it was, and be fixed later.

@hugueskamba is that correct? Can we revert this one?

hugueskamba replied:
As mentioned by @0xc0170 ,
This is a consolidation of what is upstream.
Please see this PR that reverts the consolidation: https://github.com/ARMmbed/mbed-os/pull/13590

donatieng reviewed (dismissed):
Great stuff and really glad to seeing this coming to master! I'd like to see more context in the documentation as detailed below - basically the background/why we're doing this and what's the plan.

0xc0170 committed:
CMake: Fix GCC_ARM pre-linking stage (#13575)

Make it generic for all targets. It was previously hardcoded for K64F

jeromecoutant commented on targets/TARGET_STM/TARGET_STM32F4/TARGET_STM32F439xI/TOOLCHAIN_ARM_STD/startup_stm32f439xx.S:
```diff
@@ -0,0 +1,415 @@
+;******************** (C) COPYRIGHT 2017 STMicroelectronics ********************
+;* File Name          : startup_stm32f439xx.s
```
I think this file addition is not related to this PR :-)

hugueskamba replied:
No, it was not. I see that the directory `TOOLCHAIN_ARM_STD` was renamed `TOOLCHAIN_ARM` on master. I will fix that.
Thanks

hugueskamba replied:
[Done](https://github.com/ARMmbed/mbed-os/pull/13566/commits/c0584531277fa2fa4a3293fe630c25cf5aab1628).

hugueskamba committed:
CMake: Revert mbed-trace CMake input source file to match upstream repository (#13590)

As there is currently no mbed-trace library, statements related to that
target have been commented out.

hugueskamba committed:
CMake: Remove TOOLCHAIN_ARM_STD dir for TARGET_STM32F439xI

hugueskamba committed:
CMake: Fix build failure due to BLE dir changes

hugueskamba committed:
CMake: Fix ARM toolchain configuration

mbed_generate_gcc_options_for_linker is to
be called only when GCC_ARM is selected.

hugueskamba committed:
CMake: Fix DISCO_L475VG_IOT01A ARM toolchain support

Correct location of startup code and linker file.

hugueskamba committed:
CMake: Add background and future plans to README (#13589)

hugueskamba committed:
CMake: Move modules under tools subdirectory

hugueskamba committed:
CMake: Add support for printf lib selection

hugueskamba committed:
CMake: Add support for C lib selection

Configure the selected toolchain for the selected C library

0xc0170 committed:
CMake design: update to the latest version of feature-cmake branch (#13614)

We made changes like tools/cmake folder change, removed app.cmake.

hugueskamba committed:
CMake: Fix location of M33 exception handler source file

hugueskamba committed:
CMake: Fix TFM targets build

Add linking time pre-processor macro DOMAIN_NS
for TFM targets and TrustZone non secure targets.

hugueskamba committed:
CMake: Add support for NUCLEO_F401RE

rajkan01 committed:
CMake: Add a function to generate executable artifacts with memap (#13695)

- Add linker option to generate map file for GCC_ARM and ARM toolchain
- Add mbed_generate_map_file function to call memap.py (parse the map
file).
- Add mbed_generate_executable function to generate executable artifacts with a memory map table

hugueskamba committed:
CMake: Componentize Mbed OS into multiple CMake targets (#13732)

Aside from the core mbed-os CMake target, a number of targets have been created so they can optionally be included by application executables that require them using `target_link_libraries()`.

Co-authored-by: Martin Kojtal <martin.kojtal@arm.com>
Co-authored-by: Rajkumar Kanagaraj <rajkumar.kanagaraj@arm.com>

LDong-Arm commented on tools/cmake/toolchains/GCC_ARM.cmake:
```diff
@@ -0,0 +1,128 @@
+
+    set(_linker_options
+        "$<TARGET_PROPERTY:${target},LINK_OPTIONS>"
+    )
```
@0xc0170 
`LINK_OPTIONS` are mostly flags for linking when we trigger `arm-none-eabi-ld`). Are they relevant to _preprocessing_ which is basically just text manipulating?

Even if we do need link options, we haven't done any transformation. For compile definitions we join the list with ` -D`. But if we don't join the list, the output file `linker_options.txt` is semicolon-separated instead of space-separated.

0xc0170 replied:
>LINK_OPTIONS are mostly flags for linking when we trigger arm-none-eabi-ld). Are they relevant to preprocessing which is basically just text manipulating?


```
    # Add linking time preprocessor macro for TFM targets
    if("TFM" IN_LIST MBED_TARGET_LABELS)
        list(APPEND link_options
            "-DDOMAIN_NS=1"
        )
    endif()
```

true, we do not need most of the options but if some component adds a define like this, we would skip it. Or is there a better way?

> Even if we do need link options, we haven't done any transformation. For compile definitions we join the list with -D. But if we don't join the list, the output file linker_options.txt is semicolon-separated instead of space-separated.

Do we need to? Macros must be added to link options as `-Dxxxx`, therefore we just pass this to cpp to preprocess.

LDong-Arm commented on tools/cmake/toolchains/GCC_ARM.cmake:
```diff
@@ -0,0 +1,128 @@
+        "$<TARGET_PROPERTY:${target},LINK_OPTIONS>"
+    )
+
+    set(_compile_definitions 
+        "$<$<BOOL:${_compile_definitions}>:-D$<JOIN:${_compile_definitions}, -D>>"
+    )
```
I'm not sure this can properly handle macros whose values contain spaces? Does the generated `compile_time_defs.txt` look good for you?

0xc0170 replied:
@hugueskamba would you be able to verify? I am having issues with the latest tools and symantec (waiting for the update to allow me to use mbed-tools again). From what I recall, we have lot of config values that have spaces - like array values in macros, haven't seen any toolchain error so far

rajkan01 committed:
CMake: replace usage of the mbed_add_cmake_directory_if_labels() function (#13754)

Directories that start with special prefixes (TARGET_, FEATURE_, COMPONENT_)  are added to the build based on Mbed target configuration from targets.json instead of calling utility function mbed_add_cmake_directory_if_labels().

0xc0170 committed:
cmake: fix present inclusion

We use only _PRESENT macros for components if they are enabled. Use the same in
the mbed.h file or anywhere else.

0xc0170 committed:
CMake: fix for Gcc Arm preprocessing linker file

A linker script needs symbols (stack size, app size, etc). They are basic values or defines. Not any array like or string like macros. We should filter these, as they are not valid anyway.

The other option to fix this would be to fix all the macros but I dont think it is needed as these config values won't be used in the linker script anyway.

Not allowed in ld files macros with spaces, like MACRO={0, 2, 3} or MACRO=(4 * 2000).

hugueskamba committed:
CMake: Replace tested BLE application in CI description

BLE_Button, which was the previously tested BLE example, is
being removed in mbed-os-example-ble. Replace it with BLE_LED.

0xc0170 committed:
CMake: Fix travis fetch PR (#13763)

Either fetch TRAVIS_COMMIT or the entire pull request (depending on the TRAVIS_PULL_REQUEST environment variable).

hugueskamba committed:
CMake: Fix BLE private address linker error

List newly added files in CMake input source files.

0xc0170 commented on CMakeLists.txt:
```diff
@@ -0,0 +1,198 @@
+# TODO CMake: This component is made visible here so its source files in
+# drivers/ can be added and it can be linked against by libraries in storage/.
+# Should the source files be moved from drivers/ to storage/ ?
+add_library(mbed-os-device_key INTERFACE)
```
Is this tracked via internal ticket so we can fix it ?

0xc0170 reviewed (commented):
Otherwise looks fine to me, still more to do but we should get this to master as soon as possible to start working from the main branch.

0xc0170 commented on targets/TARGET_STM/USBPhyHw.h:
```diff
@@ -27,6 +27,7 @@
 #include "platform/mbed_power_mgmt.h"
 #include "USBPhy.h"
 #include "PeripheralPins.h"
+#include "string.h"
```
We need to revert this one, it's on master already (proper fix, we needed more headers)

0xc0170 replied:
Fixed via https://github.com/ARMmbed/mbed-os/pull/13773

0xc0170 replied:
fixed now here

0xc0170 reviewed (commented):
I'll send a fix. I dont see any file change or config, dont understand why networking examples are failing

rajkan01 committed:
CMake: split rtos from core target

Add or remove and directories containing source files, header files, and macro definitions that implement RTOS support.
This includes:
- cmsis/CMSIS_5/CMSIS/RTOS2/
- cmsis/device/rtos/
- rtos/source/ConditionVariable.cpp
- rtos/source/Thread.cpp

0xc0170 committed:
CMake: remove string.h in STM32 usb, already included (#13773)

hugueskamba committed:
CMake: Restore rtos api and netsocket presence macro names

The old tools rely on the old names to build

jamesbeyond reviewed (commented):
Hi, about testing part. the CI for this PR is green, but the current CI not contains Jenkins cmake build test.  We made some attempt to enabled cmake build tests on examples in : https://github.com/ARMmbed/mbed-os/pull/13477#issuecomment-712242702
it shows `DISCO_L475VG_IOT01A` failed to build  `mbed-os-example-devicekey` and  `mbed-os-example-nfc/NFC_SmartPoster`  with cmake, all rest supported examples seems OK.
If those 2 issues can be fixed, we can enabled cmake test in the Jenkins pipeline straight forward

rajkan01 committed:
Update CMake based on new STM32L4 directory structure

0xc0170 committed:
CMake: fix application config

These settings like enable_language should be done in the application and just once.
We hit the issue when you expose sources to an app (interface or public), CMake errors as some of the internal settings have not been configured.

0xc0170 committed:
CMake: remove TODO as it is not valid anymore

hugueskamba committed:
CMake: Rename CMake targets

* mbed-os renamed mbed-core
* mbed-os-<COMPONENT> renamed mbed-<COMPONENT>

hugueskamba committed:
CMake: Create a mbed-os alias library

This ensures that applications which link with the mbed-os
CMake target still work. We want applications that require the OS
to use the mbed-os CMake target as we will be creating a new one
that will include mbed-core and mbed-rtos.

Commit:
TEST: update example scripts for cmake

rajkan01 committed:
CMake: add mbed-os and mbed-baremetal targets

mbed-os consists of mbed-core and mbed-rtos
mbed-baremetal consists of mbed-core

The main change is for mbed-core. Changing from object library to be interface. This way it allows us to do the above to have 2 main targets for users to use.

This should be backward compatible change as mbed-os target we used contains the same files/options as previously set.

0xc0170 committed:
Update rtos/CMakeLists.txt

Co-authored-by: Hugues Kamba <41612201+hugueskamba@users.noreply.github.com>

0xc0170 committed:
CMake: fix the output bin-hex file

$<TARGET_FILE:${target}> is the output from CMake. As it can contain suffix for some toolchains, we can use ${target} instead (output from hex/bin generation).

0xc0170 committed:
Update CMakeLists.txt

Co-authored-by: Rajkumar Kanagaraj <45570536+rajkan01@users.noreply.github.com>

hugueskamba committed:
CMake: Add support for NUCLEO_F303K8

Added the board so it can be used for testing the bare-metal profile on a Mbed 2 board

Commit:
TEST: remove cmake build test from TravisCI

ladislas committed:
Activate ccache for CMake

Commit:
TEST: add BLE example to cmake example json

0xc0170 committed:
CMake: ninja reponse files disabled for ARMClang on Windows

Issue https://gitlab.kitware.com/cmake/cmake/-/issues/21093

Until this is fixed, we should disable response files for ARMClang + ninja on Windows.
Other toolchains and host systems should benefit from response files thus not disabling them
completely.

This fixes the issue with not finding includes. It's not trivial to find the root cause, it took me a while to figure out why ARMClang can't find the paths.

I moved the check to Mbed OS main cmake. It should not be in the toolchain file as it is not
related to the toolchain but to generator. We need toolchain properly set up first.

Note, I had to protect setting CMAKE_NINJA_FORCE_RESPONSE_FILE. If I set it to 0, ninja
would still use rsp files so I rather protected it and define that variable only when
required, not always.

Co-authored-by: Hugues Kamba <41612201+hugueskamba@users.noreply.github.com>

Commit:
TEST: fix a bug on example scripts

Commit:
restrict BLE example only build BLE_LED

hugueskamba committed:
CMake: Create separate function to set MCU core compile defintions

The new function (mbed_set_cpu_core_definitions()) should always be
called as it has defintions needed by Mbed OS to work with that
particular MCU core.
mbed_set_cpu_core_options() should only be called if a third-party
toolchain file has not been used as it has compile options that can
cause conflicts.

hugueskamba committed:
CMake: Set compile defintions for supported toolchain at top level input source file

The compile defintions should always be added depending on the
selected toolchain.
mbed_set_toolchain_options() should only be called if a third-party
toolchain file has not been used as it has compile options that can
cause conflicts.

hugueskamba committed:
CMake: Fix third party toolchain file suppport

Compile defintions are always included. However, functions that set compile options
are only called if a third party toolchain file is not used

rajkan01 committed:
Replace PUBLIC to INTERFACE keyword

0xc0170 committed:
CMake: 3.19.0 update to fix ARMClang ASM defines

0xc0170 committed:
CMake readme: update mbed-tools to 3.4

Commit:
TEST: use mbed-tools build wrapper replace cmake build commands

0xc0170 committed:
CMake nrf52480: fix startup/linker inclusion

0xc0170 commented on targets/TARGET_NORDIC/TARGET_NRF5x/TARGET_NRF52/TARGET_MCU_NRF52840/device/CMakeLists.txt:
```diff
@@ -0,0 +1,38 @@
+        set(STARTUP_FILE TOOLCHAIN_IAR/startup_NRF52840_IAR.S)
+    endif()
+    target_sources(mbed-core INTERFACE ${STARTUP_FILE})
+endfunction()
```
I've missed this one (@hugueskamba you pinpointed it but my search failed anyway). I'll send PR now to fix this (not blocking this PR anyway).

donatieng commented on CMakeLists.txt:
```diff
@@ -0,0 +1,211 @@
+include(${MBED_ROOT}/tools/cmake/core.cmake)
+include(${MBED_ROOT}/tools/cmake/profile.cmake)
+
+add_library(mbed-core INTERFACE)
```
👍

donatieng commented on CMakeLists.txt:
```diff
@@ -0,0 +1,211 @@
+
+add_library(mbed-core INTERFACE)
+
+add_library(mbed-os INTERFACE)
```
👍

donatieng reviewed (dismissed):
Let's go! 😉

Patater commented on CMakeLists.txt:
```diff
@@ -0,0 +1,211 @@
+
+# This is the boilerplate for Mbed OS
+
+cmake_minimum_required(VERSION 3.19.0 FATAL_ERROR)
```
Could we add an item on our backlog to only require this version if armclang is being used? I don't believe gcc needs this new of a CMake verison.

hugueskamba replied:
Thanks.
A ticket has been added to the backlog.

Patater replied:
Could you link me to the ticket please? I couldn't find after a GitHub search. Thanks.

rajkan01 committed:
CMake: Remove depth level argument to match map console prints the same depth

and format as old mbed-CLI

rwalton-arm commented on CMakeLists.txt:
```diff
@@ -0,0 +1,211 @@
+cmake_minimum_required(VERSION 3.19.0 FATAL_ERROR)
+
+# Using relative paths behavior
+if(POLICY CMP0076)
```
Do we need to set this policy if we force the cmake version to `3.19.0`?

0xc0170 replied:
I'll create a ticket, we got some relative paths and somewhere using source dir variable. I recall we added this long time ago to suppress Cmake warning. We'll test with the current code base, to check if this can be removed.

0xc0170 replied:
IOTOSM-2643 created

rwalton-arm commented on CMakeLists.txt:
```diff
@@ -0,0 +1,211 @@
+mbed_set_profile_options(mbed-core ${MBED_TOOLCHAIN})
+
+set(MBED_TARGET_LABELS
+         ${MBED_TARGET_LABELS} CACHE INTERNAL ""
```
This indentation looks inconsistent. Also, why do we set `MBED_TARGET_LABELS` to the value of `MBED_TARGET_LABELS`?

rajkan01 replied:
Addressed in PR #13870

rwalton-arm commented on CMakeLists.txt:
```diff
@@ -0,0 +1,211 @@
+# Configures the application
+#
+function(mbed_configure_app_target target)
+    mbed_set_language_standard(${target})
```
Why can't we just use `target_compile_features(cxx_std_xx)` here?

rajkan01 replied:
Addressed in PR #13870

rwalton-arm commented on CMakeLists.txt:
```diff
@@ -0,0 +1,211 @@
+         TARGET
+             ${target}
+         POST_BUILD
+             COMMAND ${Python3_EXECUTABLE} ${MBED_ROOT}/tools/memap.py -t ${MBED_TOOLCHAIN} -d 4 ${CMAKE_BINARY_DIR}/${target}${CMAKE_EXECUTABLE_SUFFIX}.map
```
Do we need to add memory map generation to the new tools, or is there another off the shelf tool we can use for it? IIUC at some point we're going to remove the old tools and this will stop working.

0xc0170 replied:
I know about elf tools but haven't yet seen something similar to this. 

Some scripts will be needed (not yet ported to CMake, this is one of them. Others I can think of: targets post hooks, psa (tfm), we have other scripts for tests, imporint cmsis, etc). We soon will have time to look at these so prior that, we will talk to you to look at what we should port and how.

rwalton-arm commented on connectivity/drivers/ble/FEATURE_BLE/TARGET_Ambiq_Micro/CMakelists.txt:
```diff
@@ -0,0 +1,16 @@
+# Copyright (c) 2020 ARM Limited. All rights reserved.
```
I think this file should be named `CMakeLists.txt`.

rajkan01 replied:
Addressed in PR #13870

rwalton-arm commented on connectivity/drivers/ble/FEATURE_BLE/TARGET_Ambiq_Micro/TARGET_Apollo3/CMakelists.txt:
```diff
@@ -0,0 +1,17 @@
+# Copyright (c) 2020 ARM Limited. All rights reserved.
```
I think this file should be named `CMakeLists.txt`.

rajkan01 replied:
Addressed in PR #13870

rwalton-arm commented on connectivity/drivers/ble/FEATURE_BLE/TARGET_NORDIC/TARGET_NORDIC_CORDIO/TARGET_MCU_NRF52840/CMakeLists.txt:
```diff
@@ -0,0 +1,2 @@
+# Copyright (c) 2020 ARM Limited. All rights reserved.
```
Why do we need this empty `CMakeLists.txt`?

rajkan01 replied:
Removed the empty CMakeLists.txt file

rajkan01 replied:
Addressed in PR #13870

rwalton-arm commented on connectivity/drivers/mbedtls/FEATURE_CRYPTOCELL310/TARGET_MCU_NRF52840/CMakeLists.txt:
```diff
@@ -0,0 +1,43 @@
+# Copyright (c) 2020 ARM Limited. All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
+
+function(_mbed_get_libcc_310_ext)
```
Do we need to wrap this library selection logic in functions? Are the functions being used outside of this file?

rajkan01 replied:
It is not used anywhere, I cleaned up

rajkan01 replied:
Addressed in PR #13870

rwalton-arm commented on connectivity/drivers/mbedtls/FEATURE_CRYPTOCELL310/TARGET_MCU_NRF52840/CMakeLists.txt:
```diff
@@ -0,0 +1,43 @@
+target_sources(mbed-mbedtls-cryptocell310
+    INTERFACE
+        crypto_device_platform.c
+)
```
Missing newline.

rajkan01 replied:
Addressed in PR #13870

rwalton-arm commented on connectivity/drivers/mbedtls/FEATURE_CRYPTOCELL310/binaries/CMakeLists.txt:
```diff
@@ -0,0 +1,18 @@
+    )
+endfunction()
+
+_mbed_get_libcc_310_core()
```
As above; seems unnecessary to make this into a function if we only call it one place in the file where it's defined.

rajkan01 replied:
Addressed in PR #13870

rwalton-arm commented on connectivity/nanostack/sal-stack-nanostack/source/DHCPv6_Server/CMakeLists.txt:
```diff
@@ -0,0 +1,12 @@
+
+target_include_directories(mbed-nanostack-sal_stack
+    INTERFACE
+        ${CMAKE_CURRENT_SOURCE_DIR}
```
Is there a reason we use `CMAKE_CURRENT_SOURCE_DIR` here, and below, when we use `.` everywhere else? Should we try to be consistent?

0xc0170 replied:
Related to relative paths, we got internal ticket to fix this.

rwalton-arm commented on connectivity/netsocket/CMakeLists.txt:
```diff
@@ -0,0 +1,66 @@
+# Copyright (c) 2020 ARM Limited. All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
+
+# TODO CMake: Perhaps move this/these file(s)
```
Where do you propose we move them to? Could you explain in the `TODO`?

rajkan01 replied:
Addressed in PR #13870

rwalton-arm commented on platform/mbed-trace/source/CMakeLists.txt:
```diff
@@ -1,18 +1,18 @@
-    add_library( mbed-trace
-        mbed_trace.c
-    )
+    # add_library( mbed-trace
```
Would it be worth explaining, in a comment, why all these directives are commented out?

rwalton-arm reviewed (commented):
Looks good. I didn't see anything that should block the PR, only a couple of minor comments/queries.

hugueskamba committed:
CMake: Rename Mbed OS path variable for consistency

Rename the CMake variable expected from applications from MBED_ROOT to
MBED_PATH to be consistent with the naming used for the configuration file
(MBED_CONFIG_PATH) also expected from applications.

hugueskamba committed:
CMake: Fix Microlib selection (#13871)

As the assembler does not seem to recognise the request to the
linker to use microlib, remove that instruction.
Microlib is still used as the linker selects microlib as a result
of being explicitly asked to do so.

Patater reviewed (dismissed):
I see some merge commits that need to be rebased out before this is merged.

Otherwise, OK. Good starting point.

rajkan01 committed:
CMake: Validate selected application profile (#13856)

Check if the selected application profile is supported by the selected Mbed Target
